### PR TITLE
GLL/GL on simplices with recursive points

### PR DESCRIPTION
**Pull request by pbrubeck**

### CODE REVIEW

File: FIAT/recursive_points.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    imax = isum - (d - 1) * imin
+    if imax < imin:
+        return
+    for i in range(imin, imax):
+        for a in multiindex_equal(d - 1, isum - i, imin=imin):
+            yield a + (i,)
+    yield (imin,) * (d - 1) + (imax,)
+
+
+class RecursivePointSet(object):
```
Comment (rckirby) [inline FIAT/recursive_points.py:39]: Do we want to re-implement Toby's work here or just create a dependency?  If we did, we would have access to all the other point sets he implements without re-doing them ourselves.

File: FIAT/recursive_points.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                ref_el.get_vertices_of_subcomplex(
+                    ref_el.get_topology()[dim][entity_id])
+            return self.recursive_points(entity_verts, order, 1)
+        elif dim == ref_el.get_spatial_dimension():
+            return self.recursive_points(ref_el.get_vertices(), order, 1)
+        else:
+            raise ValueError("illegal dimension")
+
+
+if __name__ == "__main__":
```
Comment (rckirby) [inline FIAT/recursive_points.py:106]: Should we consider moving this code into a demo/test/somewhere besides a module file?

**Review by rckirby — COMMENTED** [review]
This all seems eminently reasonable.  I had a few comments.  Major point is whether we want to re-implement `recursivenodes` or create a dependency.  It should be a simple install procedure and would give us more features and let us incrementally get rid of old code like for orthogonal bases!  But before we embark on a major undertaking like that, we should probably see where elements, etc, wind up with ongoing UFL work.

**Review by rckirby — APPROVED** [review]
I still think the discussion of using vs recreating recursivenodes is worth having, but it's probably best not to create a new dependency at this point?

File: FIAT/gauss_lobatto_legendre.py [hunk 3]
```diff
@@ -8,38 +8,69 @@
...
+        if ref_el.shape not in {LINE, TRIANGLE, TETRAHEDRON}:
+            raise ValueError("Gauss-Lobatto-Legendre elements are only defined on simplices.")
         dual = GaussLobattoLegendreDualSet(ref_el, degree)
-        points = []
-        for node in dual.nodes:
-            # Assert singleton point for each node.
-            pt, = node.get_point_dict().keys()
-            points.append(pt)
-        poly_set = LagrangePolynomialSet(ref_el, points)
+        if ref_el.shape == LINE:
```
Comment (dham) [inline FIAT/gauss_lobatto_legendre.py:66]: comment why we're doing this.

File: test/unit/test_gauss_legendre.py [hunk 4]
```diff
@@ -23,25 +23,64 @@
...
+    quadrature_points = lr.pts
+
+    entity_dofs = fe.entity_dofs()
+    edge_dofs = entity_dofs[1]
+    for entity in edge_dofs:
+        if len(edge_dofs[entity]) > 0:
+            transform = s.get_entity_transform(1, entity)
+            assert np.allclose(points[edge_dofs[entity]], np.array(list(map(transform, quadrature_points))))
+
+    # TODO add rotational symmetry tests
```
Comment (dham) [inline test/unit/test_gauss_legendre.py:81]: remove

**Review by dham — APPROVED** [review]

**Comment by tisaac** [comment]
> I still think the discussion of using vs recreating recursivenodes is worth having, but it's probably best not to create a new dependency at this point?

You can compare the approaches in other projects.

- The construction is intentionally simple and I had expected reimplementation to be the most common choice.  This is what [basix does](https://github.com/inducer/meshmode/blob/9e942e6f602a21d433ffbae76f6c4b3089166e4f/setup.py).

- At Andreas Klöckner's prodding, I kept the transitive requirements lightweight -- only numpy is required to generate the nodes -- so he has `recursivenodes` as a dependency in [meshmode](https://github.com/inducer/meshmode/blob/9e942e6f602a21d433ffbae76f6c4b3089166e4f/setup.py#L51).  The package is on an MIT license.  I have not has reason to test the package with the latest python (its minimum requirement is 3.6), but am happy to address issues if they come up.

**Pull request merged**